README.md
# stomp_ros

[![license - apache 2.0](https://img.shields.io/:license-Apache%202.0-yellowgreen.svg)](https://opensource.org/licenses/Apache-2.0)

## Build Status

Platform             | CI Status
---------------------|:---------
Linux (Focal)        |[![Build Status](https://github.com/ros-industrial/stomp/actions/workflows/focal_build.yml/badge.svg?branch=main)](https://github.com/ros-industrial/stomp/actions/workflows/focal_build.yml)
Linux (Bionic)       |[![Build Status](https://github.com/ros-industrial/stomp/actions/workflows/bionic_build.yml/badge.svg?branch=main)](https://github.com/ros-industrial/stomp/actions/workflows/bionic_build.yml)
Windows              |[![Build Status](https://github.com/ros-industrial/stomp/actions/workflows/windows_noetic_build.yml/badge.svg?branch=main)](https://github.com/ros-industrial/stomp/actions/workflows/windows_noetic_build.yml)
Lint  (Clang-Format) |[![Build Status](https://github.com/ros-industrial/stomp/actions/workflows/clang_format.yml/badge.svg?branch=main)](https://github.com/ros-industrial/stomp/actions/workflows/clang_format.yml)
Lint  (CMake-Format) |[![Build Status](https://github.com/ros-industrial/stomp/actions/workflows/cmake_format.yml/badge.svg?branch=main)](https://github.com/ros-industrial/stomp/actions/workflows/cmake_format.yml)

#### Build
- Build the workspace:
  - Cd into the catkin workspace directory and type the following command:
```
catkin build
```

#### Build Unit Test
Cd into the catkin workspace directory and type the following command:
```
catkin build --cmake-args -DSTOMP_ENABLE_TESTING=ON 
```

#### Run Unit Test
Cd into the catkin workspace stomp build directory and type the following command:
```
ctest
```


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package stomp
^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.1.2 (2023-02-17)
------------------

0.1.1 (2017-03-24)
------------------

0.1.0 (2017-03-14)
------------------
* Initial release
* Contributors: Levi H Armstrong, Jorge Nicho


test/utest.cpp
/**
 * @file utest.cpp
 * @brief This executes the gtest code for stomp
 *
 * @author Jorge Nicho
 * @date March 25, 2016
 * @version TODO
 * @bug No known bugs
 *
 * @copyright Copyright (c) 2016, Southwest Research Institute
 *
 * @par License
 * Software License Agreement (Apache License)
 * @par
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * @par
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <gtest/gtest.h>

/** @brief This executes all tests for the stomp package */
int main(int argc, char** argv)
{
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}


test/stomp_3dof.cpp
/**
 * @file stomp_3dof.cpp
 * @brief This contains gtest code for stomp
 *
 * @author Jorge Nicho
 * @date March 25, 2016
 * @version TODO
 * @bug No known bugs
 *
 * @copyright Copyright (c) 2016, Southwest Research Institute
 *
 * @par License
 * Software License Agreement (Apache License)
 * @par
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * @par
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <iostream>
#include <Eigen/Dense>
#include <gtest/gtest.h>
#include <stomp/stomp.h>
#include <stomp/task.h>

using Trajectory = Eigen::MatrixXd; /**< Assign Type Trajectory to Eigen::MatrixXd Type */

const std::size_t NUM_DIMENSIONS = 3;                               /**< Number of parameters to optimize */
const std::size_t NUM_TIMESTEPS = 20;                               /**< Number of timesteps */
const double DELTA_T = 0.1;                                         /**< Timestep in seconds */
const std::vector<double> START_POS = { 1.4, 1.4, 0.5 };            /**< Trajectory starting position */
const std::vector<double> END_POS = { -1.25, 1.0, -0.26 };          /**< Trajectory ending posiiton */
const std::vector<double> BIAS_THRESHOLD = { 0.050, 0.050, 0.050 }; /**< Threshold to determine whether two trajectories
                                                                       are equal */
const std::vector<double> STD_DEV = { 1.0, 1.0, 1.0 }; /**< Standard deviation used for generating noisy parameters */

using namespace stomp;

/** @brief A dummy task for testing STOMP */
class DummyTask : public Task
{
public:
  /**
   * @brief A dummy task for testing Stomp
   * @param parameters_bias default parameter bias used for computing cost for the test
   * @param bias_thresholds threshold to determine whether two trajectories are equal
   * @param std_dev standard deviation used for generating noisy parameters
   */
  DummyTask(const Trajectory& parameters_bias,
            const std::vector<double>& bias_thresholds,
            const std::vector<double>& std_dev)
    : parameters_bias_(parameters_bias), bias_thresholds_(bias_thresholds), std_dev_(std_dev)
  {
    // generate smoothing matrix
    int num_timesteps = parameters_bias.cols();
    generateSmoothingMatrix(num_timesteps, 1.0, smoothing_M_);
    srand(1);
  }

  /** @brief See base clase for documentation */
  bool generateNoisyParameters(const Eigen::MatrixXd& parameters,
                               std::size_t start_timestep,
                               std::size_t num_timesteps,
                               int iteration_number,
                               int rollout_number,
                               Eigen::MatrixXd& parameters_noise,
                               Eigen::MatrixXd& noise) override
  {
    double rand_noise;
    for (std::size_t d = 0; d < parameters.rows(); d++)
    {
      for (std::size_t t = 0; t < parameters.cols(); t++)
      {
        rand_noise = static_cast<double>(rand() % RAND_MAX) / static_cast<double>(RAND_MAX - 1);  // 0 to 1
        rand_noise = 2 * (0.5 - rand_noise);
        noise(d, t) = rand_noise * std_dev_[d];
      }
    }

    parameters_noise = parameters + noise;

    return true;
  }

  bool computeCosts(const Trajectory& parameters,
                    std::size_t start_timestep,
                    std::size_t num_timesteps,
                    int iteration_number,
                    Eigen::VectorXd& costs,
                    bool& validity) override
  {
    return computeNoisyCosts(parameters, start_timestep, num_timesteps, iteration_number, -1, costs, validity);
  }

  bool computeNoisyCosts(const Trajectory& parameters,
                         std::size_t start_timestep,
                         std::size_t num_timesteps,
                         int iteration_number,
                         int rollout_number,
                         Eigen::VectorXd& costs,
                         bool& validity) override
  {
    costs.setZero(num_timesteps);
    double diff;
    double cost = 0.0;
    validity = true;

    for (std::size_t t = 0u; t < num_timesteps; t++)
    {
      cost = 0;
      for (std::size_t d = 0u; d < parameters.rows(); d++)
      {
        diff = std::abs(parameters(d, t) - parameters_bias_(d, t));
        if (diff > std::abs(bias_thresholds_[d]))
        {
          cost += diff;
          validity = false;
        }
      }

      costs(t) = cost;
    }

    return true;
  }

  bool filterParameterUpdates(std::size_t start_timestep,
                              std::size_t num_timesteps,
                              int iteration_number,
                              const Eigen::MatrixXd& parameters,
                              Eigen::MatrixXd& updates) override
  {
    return smoothParameterUpdates(start_timestep, num_timesteps, iteration_number, updates);
  }

protected:
  /**
   * @brief Perform a smooth update given a noisy update
   * @param start_timestep starting timestep
   * @param num_timesteps number of timesteps
   * @param iteration_number number of interations allowed
   * @param updates returned smooth update
   * @return True if successful, otherwise false
   */
  bool smoothParameterUpdates(std::size_t start_timestep,
                              std::size_t num_timesteps,
                              int iteration_number,
                              Eigen::MatrixXd& updates)
  {
    for (auto d = 0u; d < updates.rows(); d++)
    {
      updates.row(d).transpose() = smoothing_M_ * (updates.row(d).transpose());
    }

    return true;
  }

protected:
  Trajectory parameters_bias_;          /**< Parameter bias used for computing cost for the test */
  std::vector<double> bias_thresholds_; /**< Threshold to determine whether two trajectories are equal */
  std::vector<double> std_dev_;         /**< Standard deviation used for generating noisy parameters */
  Eigen::MatrixXd smoothing_M_;         /**< Matrix used for smoothing the trajectory */
};

/**
 * @brief Compares whether two trajectories are close to each other within a threshold.
 * @param optimized optimized trajectory
 * @param desired desired trajectory
 * @param thresholds used to determine if two values are equal
 * @return True if the difference between the two is less than the threshold, otherwise false
 */
bool compareDiff(const Trajectory& optimized, const Trajectory& desired, const std::vector<double>& thresholds)
{
  auto num_dimensions = optimized.rows();
  Trajectory diff = Trajectory::Zero(num_dimensions, optimized.cols());
  for (auto d = 0u; d < num_dimensions; d++)
  {
    diff.row(d) = optimized.row(d) - desired.row(d);
    diff.row(d).cwiseAbs();
    if ((diff.row(d).array() > thresholds[d]).any())
    {
      return false;
    }
  }

  return true;
}

/**
 * @brief Create the STOMP configuration for the 3 DOF problem.
 * @return  StompConfiguration
 */
StompConfiguration create3DOFConfiguration()
{
  StompConfiguration c;
  c.num_timesteps = NUM_TIMESTEPS;
  c.num_iterations = 40;
  c.num_dimensions = NUM_DIMENSIONS;
  c.delta_t = DELTA_T;
  c.control_cost_weight = 0.0;
  c.initialization_method = TrajectoryInitializations::LINEAR_INTERPOLATION;
  c.num_iterations_after_valid = 0;
  c.num_rollouts = 20;
  c.max_rollouts = 20;

  return c;
}

/**
 * @brief Compute a linear interpolated trajectory given a start and end state
 * @param start start position
 * @param end last position
 * @param num_timesteps number of timesteps
 * @param traj returned linear interpolated trajectory
 */
void interpolate(const std::vector<double>& start,
                 const std::vector<double>& end,
                 std::size_t num_timesteps,
                 Trajectory& traj)
{
  auto dimensions = start.size();
  traj = Eigen::MatrixXd::Zero(dimensions, num_timesteps);
  for (auto d = 0u; d < dimensions; d++)
  {
    double delta = (end[d] - start[d]) / (num_timesteps - 1);
    for (auto t = 0u; t < num_timesteps; t++)
    {
      traj(d, t) = start[d] + t * delta;
    }
  }
}

/** @brief This tests the Stomp constructor */
TEST(Stomp3DOF, construction)
{
  Trajectory trajectory_bias;
  interpolate(START_POS, END_POS, NUM_TIMESTEPS, trajectory_bias);
  TaskPtr task(new DummyTask(trajectory_bias, BIAS_THRESHOLD, STD_DEV));

  Stomp stomp(create3DOFConfiguration(), task);
}

/** @brief This tests the Stomp solve method for a default case */
TEST(Stomp3DOF, solve_default)
{
  Trajectory trajectory_bias;
  interpolate(START_POS, END_POS, NUM_TIMESTEPS, trajectory_bias);
  TaskPtr task(new DummyTask(trajectory_bias, BIAS_THRESHOLD, STD_DEV));

  StompConfiguration config = create3DOFConfiguration();
  Stomp stomp(config, task);

  Trajectory optimized;
  stomp.solve(START_POS, END_POS, optimized);

  EXPECT_EQ(optimized.rows(), NUM_DIMENSIONS);
  EXPECT_EQ(optimized.cols(), NUM_TIMESTEPS);
  EXPECT_TRUE(compareDiff(optimized, trajectory_bias, BIAS_THRESHOLD));

  // calculate difference
  Trajectory diff;
  diff = trajectory_bias - optimized;

  std::string line_separator = "\n------------------------------------------------------\n";
  std::cout << line_separator;
  std::cout << stomp::toString(trajectory_bias);
  std::cout << line_separator;
  std::cout << toString(optimized) << "\n";
  std::cout << "Differences"
            << "\n"
            << toString(diff) << line_separator;
}

/** @brief This tests the Stomp solve method using a linear interpolated inital trajectory */
TEST(Stomp3DOF, solve_interpolated_initial)
{
  Trajectory trajectory_bias;
  interpolate(START_POS, END_POS, NUM_TIMESTEPS, trajectory_bias);
  TaskPtr task(new DummyTask(trajectory_bias, BIAS_THRESHOLD, STD_DEV));

  StompConfiguration config = create3DOFConfiguration();
  config.initialization_method = TrajectoryInitializations::LINEAR_INTERPOLATION;
  Stomp stomp(config, task);

  Trajectory optimized;
  stomp.solve(trajectory_bias, optimized);

  EXPECT_EQ(optimized.rows(), NUM_DIMENSIONS);
  EXPECT_EQ(optimized.cols(), NUM_TIMESTEPS);
  EXPECT_TRUE(compareDiff(optimized, trajectory_bias, BIAS_THRESHOLD));

  // calculate difference
  Trajectory diff;
  diff = trajectory_bias - optimized;

  std::string line_separator = "\n------------------------------------------------------\n";
  std::cout << line_separator;
  std::cout << stomp::toString(trajectory_bias);
  std::cout << line_separator;
  std::cout << toString(optimized) << "\n";
  std::cout << "Differences"
            << "\n"
            << toString(diff) << line_separator;
}

/** @brief This tests the Stomp solve method using a cubic polynomial interpolated inital trajectory */
TEST(Stomp3DOF, solve_cubic_polynomial_initial)
{
  Trajectory trajectory_bias;
  interpolate(START_POS, END_POS, NUM_TIMESTEPS, trajectory_bias);
  TaskPtr task(new DummyTask(trajectory_bias, BIAS_THRESHOLD, STD_DEV));

  StompConfiguration config = create3DOFConfiguration();
  config.initialization_method = TrajectoryInitializations::CUBIC_POLYNOMIAL_INTERPOLATION;
  Stomp stomp(config, task);

  Trajectory optimized;
  stomp.solve(trajectory_bias, optimized);

  EXPECT_EQ(optimized.rows(), NUM_DIMENSIONS);
  EXPECT_EQ(optimized.cols(), NUM_TIMESTEPS);
  EXPECT_TRUE(compareDiff(optimized, trajectory_bias, BIAS_THRESHOLD));

  // calculate difference
  Trajectory diff;
  diff = trajectory_bias - optimized;

  std::string line_separator = "\n------------------------------------------------------\n";
  std::cout << line_separator;
  std::cout << stomp::toString(trajectory_bias);
  std::cout << line_separator;
  std::cout << toString(optimized) << "\n";
  std::cout << "Differences"
            << "\n"
            << toString(diff) << line_separator;
}

/** @brief This tests the Stomp solve method using a minimum cost inital trajectory */
TEST(Stomp3DOF, solve_min_control_cost_initial)
{
  Trajectory trajectory_bias;
  interpolate(START_POS, END_POS, NUM_TIMESTEPS, trajectory_bias);
  TaskPtr task(new DummyTask(trajectory_bias, BIAS_THRESHOLD, STD_DEV));

  StompConfiguration config = create3DOFConfiguration();
  config.initialization_method = TrajectoryInitializations::MININUM_CONTROL_COST;
  Stomp stomp(config, task);

  Trajectory optimized;
  stomp.solve(trajectory_bias, optimized);

  EXPECT_EQ(optimized.rows(), NUM_DIMENSIONS);
  EXPECT_EQ(optimized.cols(), NUM_TIMESTEPS);
  EXPECT_TRUE(compareDiff(optimized, trajectory_bias, BIAS_THRESHOLD));

  // calculate difference
  Trajectory diff;
  diff = trajectory_bias - optimized;

  std::string line_separator = "\n------------------------------------------------------\n";
  std::cout << line_separator;
  std::cout << stomp::toString(trajectory_bias);
  std::cout << line_separator;
  std::cout << toString(optimized) << "\n";
  std::cout << "Differences"
            << "\n"
            << toString(diff) << line_separator;
}

/** @brief This tests Stomp solve method given 40 timesteps */
TEST(Stomp3DOF, solve_40_timesteps)
{
  Trajectory trajectory_bias;
  int num_timesteps = 40;
  interpolate(START_POS, END_POS, num_timesteps, trajectory_bias);
  TaskPtr task(new DummyTask(trajectory_bias, BIAS_THRESHOLD, STD_DEV));

  StompConfiguration config = create3DOFConfiguration();
  config.initialization_method = TrajectoryInitializations::LINEAR_INTERPOLATION;
  config.num_iterations = 100;
  config.num_timesteps = num_timesteps;
  Stomp stomp(config, task);

  Trajectory optimized;
  stomp.solve(START_POS, END_POS, optimized);

  EXPECT_EQ(optimized.rows(), NUM_DIMENSIONS);
  EXPECT_EQ(optimized.cols(), num_timesteps);
  EXPECT_TRUE(compareDiff(optimized, trajectory_bias, BIAS_THRESHOLD));

  // calculate difference
  Trajectory diff;
  diff = trajectory_bias - optimized;

  std::string line_separator = "\n------------------------------------------------------\n";
  std::cout << line_separator;
  std::cout << stomp::toString(trajectory_bias);
  std::cout << line_separator;
  std::cout << toString(optimized) << "\n";
  std::cout << "Differences"
            << "\n"
            << toString(diff) << line_separator;
}

/** @brief This tests Stomp solve method given 60 timesteps */
TEST(Stomp3DOF, solve_60_timesteps)
{
  Trajectory trajectory_bias;
  int num_timesteps = 60;
  interpolate(START_POS, END_POS, num_timesteps, trajectory_bias);
  TaskPtr task(new DummyTask(trajectory_bias, BIAS_THRESHOLD, STD_DEV));

  StompConfiguration config = create3DOFConfiguration();
  // config.initialization_method = TrajectoryInitializations::MININUM_CONTROL_COST; // this initialization method is
  // currently behaving oddly
  config.num_iterations = 100;
  config.num_timesteps = num_timesteps;
  Stomp stomp(config, task);

  Trajectory optimized;
  stomp.solve(START_POS, END_POS, optimized);

  EXPECT_EQ(optimized.rows(), NUM_DIMENSIONS);
  EXPECT_EQ(optimized.cols(), num_timesteps);
  EXPECT_TRUE(compareDiff(optimized, trajectory_bias, BIAS_THRESHOLD));

  // calculate difference
  Trajectory diff;
  diff = trajectory_bias - optimized;

  std::string line_separator = "\n------------------------------------------------------\n";
  std::cout << line_separator;
  std::cout << stomp::toString(trajectory_bias);
  std::cout << line_separator;
  std::cout << toString(optimized) << "\n";
  std::cout << "Differences"
            << "\n"
            << toString(diff) << line_separator;
}


src/utils.cpp
/**
 * @file utils.cpp
 * @brief This is a utility class for stomp
 *
 * @author Jorge Nicho
 * @date March 7, 2016
 * @version TODO
 * @bug No known bugs
 *
 * @copyright Copyright (c) 2016, Southwest Research Institute
 *
 * @par License
 * Software License Agreement (Apache License)
 * @par
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * @par
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <stomp/utils.h>
#include <cmath>
#include <iostream>
#include <Eigen/Dense>

namespace stomp
{
void generateFiniteDifferenceMatrix(int num_time_steps,
                                    DerivativeOrders::DerivativeOrder order,
                                    double dt,
                                    Eigen::MatrixXd& diff_matrix)
{
  diff_matrix = Eigen::MatrixXd::Zero(num_time_steps, num_time_steps);
  double multiplier = 1.0 / pow(dt, (int)order);
  for (int i = 0; i < num_time_steps; ++i)
  {
    for (int j = -FINITE_DIFF_RULE_LENGTH / 2; j <= FINITE_DIFF_RULE_LENGTH / 2; ++j)
    {
      int index = i + j;
      if (index < 0)
      {
        index = 0;
        continue;
      }
      if (index >= num_time_steps)
      {
        index = num_time_steps - 1;
        continue;
      }

      diff_matrix(i, index) = multiplier * FINITE_CENTRAL_DIFF_COEFFS[order][j + FINITE_DIFF_RULE_LENGTH / 2];
    }
  }
}

void generateSmoothingMatrix(int num_timesteps, double dt, Eigen::MatrixXd& projection_matrix_M)
{
  using namespace Eigen;

  // generate augmented finite differencing matrix
  int start_index_padded = FINITE_DIFF_RULE_LENGTH - 1;
  int num_timesteps_padded = num_timesteps + 2 * (FINITE_DIFF_RULE_LENGTH - 1);
  MatrixXd finite_diff_matrix_A_padded;
  generateFiniteDifferenceMatrix(
      num_timesteps_padded, DerivativeOrders::STOMP_ACCELERATION, dt, finite_diff_matrix_A_padded);

  /* computing control cost matrix (R = A_transpose * A):
   * Note: Original code multiplies the A product by the time interval.  However this is not
   * what was described in the literature
   */
  MatrixXd control_cost_matrix_R_padded = dt * finite_diff_matrix_A_padded.transpose() * finite_diff_matrix_A_padded;
  MatrixXd control_cost_matrix_R =
      control_cost_matrix_R_padded.block(start_index_padded, start_index_padded, num_timesteps, num_timesteps);
  MatrixXd inv_control_cost_matrix_R = control_cost_matrix_R.fullPivLu().inverse();

  // computing projection matrix M
  projection_matrix_M = inv_control_cost_matrix_R;
  double max = 0;
  for (auto t = 0u; t < num_timesteps; t++)
  {
    max = projection_matrix_M(t, t);
    projection_matrix_M.col(t) *=
        (1.0 / (num_timesteps * max));  // scaling such that the maximum value is 1/num_timesteps
  }
}

void differentiate(const Eigen::VectorXd& parameters,
                   DerivativeOrders::DerivativeOrder order,
                   double dt,
                   Eigen::VectorXd& derivatives)
{
  using namespace Eigen;

  derivatives = Eigen::VectorXd::Zero(parameters.size());

  // coefficient arrays
  VectorXd central_coeffs = VectorXd::Map(&FINITE_CENTRAL_DIFF_COEFFS[order][0], FINITE_DIFF_RULE_LENGTH);
  VectorXd forward_coeffs = VectorXd::Map(&FINITE_FORWARD_DIFF_COEFFS[order][0], FINITE_DIFF_RULE_LENGTH);
  VectorXd backward_coeffs = forward_coeffs.reverse();

  // check order eveness
  if (order % 2 != 0)
  {
    backward_coeffs *= -1.0;
  }

  // creating difference matrix
  int rule_length = FINITE_DIFF_RULE_LENGTH;
  int size = parameters.size();
  int skip = FINITE_DIFF_RULE_LENGTH / 2;
  Eigen::MatrixXd A = Eigen::MatrixXd::Zero(size, size);
  int start_ind, end_ind;
  for (auto i = 0; i < size; i++)
  {
    if (i < skip)
    {
      start_ind = i;
      A.row(i).segment(start_ind, rule_length) = forward_coeffs;
    }
    else if (i >= skip && i < size - skip)
    {
      start_ind = i - skip;
      A.row(i).segment(start_ind, rule_length) = central_coeffs;
    }
    else
    {
      start_ind = i - rule_length + 1;
      A.row(i).segment(start_ind, rule_length) = backward_coeffs;
    }
  }

  derivatives = A * parameters / std::pow(dt, 2);
}

void toVector(const Eigen::MatrixXd& m, std::vector<Eigen::VectorXd>& v)
{
  v.resize(m.rows(), Eigen::VectorXd::Zero(m.cols()));
  for (auto d = 0u; d < m.rows(); d++)
  {
    v[d] = m.row(d);
  }
}

std::string toString(const std::vector<Eigen::VectorXd>& data)
{
  Eigen::IOFormat clean_format(4, 0, ", ", "\n", "[", "]");
  Eigen::MatrixXd m = Eigen::MatrixXd::Zero(data.size(), data.front().size());
  std::stringstream ss;
  for (auto d = 0u; d < data.size(); d++)
  {
    m.row(d) = data[d].transpose();
  }

  ss << m.format(clean_format);
  return ss.str();
}

std::string toString(const Eigen::MatrixXd& data)
{
  Eigen::IOFormat clean_format(4, 0, ", ", "\n", "[", "]");
  std::stringstream ss;
  ss << data.format(clean_format);
  return ss.str();
}

std::string toString(const Eigen::VectorXd& data)
{
  Eigen::IOFormat clean_format(4, 0, ", ", "\n", "[", "]");
  std::stringstream ss;
  ss << data.transpose().format(clean_format);
  return ss.str();
}

}  // namespace stomp


examples/stomp_example.cpp
/**
 * @file stomp_example.cpp
 * @brief Demonstrates how to optimize a trajectory using STOMP
 *
 * @author Jorge Nicho
 * @date Dec 14, 2016
 * @version TODO
 * @bug No known bugs
 *
 * @copyright Copyright (c) 2016, Southwest Research Institute
 *
 * @par License
 * Software License Agreement (Apache License)
 * @par
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * @par
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <iostream>
#include <Eigen/Dense>
#include <stomp/stomp.h>
#include "simple_optimization_task.h"

using Trajectory = Eigen::MatrixXd; /**< Assign Type Trajectory to Eigen::MatrixXd Type */

/**< Declaring optimization variables */
static const std::size_t NUM_DIMENSIONS = 3;                               /**< Number of parameters to optimize */
static const std::size_t NUM_TIMESTEPS = 20;                               /**< Number of timesteps */
static const double DELTA_T = 0.1;                                         /**< Timestep in seconds */
static const std::vector<double> START_POS = { 1.4, 1.4, 0.5 };            /**< Trajectory starting position */
static const std::vector<double> END_POS = { -1.25, 1.0, -0.26 };          /**< Trajectory ending posiiton */
static const std::vector<double> BIAS_THRESHOLD = { 0.050, 0.050, 0.050 }; /**< Threshold to determine whether two
                                                                              trajectories are equal */
static const std::vector<double> STD_DEV = { 1.0, 1.0, 1.0 }; /**< Standard deviation used for generating noisy
                                                                 parameters */

/**
 * @brief Creates a STOMP configuration object with default parameters.
 * @return A STOMP configuration object
 */
stomp::StompConfiguration create3DOFConfiguration()
{
  //! [Create Config]
  using namespace stomp;

  StompConfiguration c;
  c.num_timesteps = NUM_TIMESTEPS;
  c.num_iterations = 40;
  c.num_dimensions = NUM_DIMENSIONS;
  c.delta_t = DELTA_T;
  c.control_cost_weight = 0.0;
  c.initialization_method = TrajectoryInitializations::LINEAR_INTERPOLATION;
  c.num_iterations_after_valid = 0;
  c.num_rollouts = 20;
  c.max_rollouts = 20;
  //! [Create Config]

  return c;
}

/**
 * @brief Compares whether two trajectories are close to each other within a threshold.
 * @param optimized optimized trajectory
 * @param desired desired trajectory
 * @param thresholds used to determine if two values are equal
 * @return True if the difference between the two is less than the threshold, otherwise false
 */
bool compareDiff(const Trajectory& optimized, const Trajectory& desired, const std::vector<double>& thresholds)
{
  auto num_dimensions = optimized.rows();
  Trajectory diff = Trajectory::Zero(num_dimensions, optimized.cols());
  for (auto d = 0u; d < num_dimensions; d++)
  {
    diff.row(d) = optimized.row(d) - desired.row(d);
    diff.row(d).cwiseAbs();
    if ((diff.row(d).array() > thresholds[d]).any())
    {
      return false;
    }
  }

  return true;
}

/**
 * @brief Compute a linear interpolated trajectory given a start and end state
 * @param start start position
 * @param end last position
 * @param num_timesteps number of timesteps
 * @param traj returned linear interpolated trajectory
 */
void interpolate(const std::vector<double>& start,
                 const std::vector<double>& end,
                 std::size_t num_timesteps,
                 Trajectory& traj)
{
  auto dimensions = start.size();
  traj = Eigen::MatrixXd::Zero(dimensions, num_timesteps);
  for (auto d = 0u; d < dimensions; d++)
  {
    double delta = (end[d] - start[d]) / (num_timesteps - 1);
    for (auto t = 0u; t < num_timesteps; t++)
    {
      traj(d, t) = start[d] + t * delta;
    }
  }
}

int main(int argc, char** argv)
{
  using namespace stomp_examples;
  using namespace stomp;

  /**< Creating a Task with a trajectory bias **/
  Trajectory trajectory_bias;
  interpolate(START_POS, END_POS, NUM_TIMESTEPS, trajectory_bias);

  //! [Create Task Object]
  TaskPtr task(new SimpleOptimizationTask(trajectory_bias, BIAS_THRESHOLD, STD_DEV));
  //! [Create Task Object]

  //! [Create STOMP]
  /**< Creating STOMP to find a trajectory close enough to the bias **/
  StompConfiguration config = create3DOFConfiguration();
  Stomp stomp(config, task);
  //! [Create STOMP]

  //! [Solve]
  /**< Optimizing a trajectory close enough to the bias is produced **/
  Trajectory optimized;
  if (stomp.solve(START_POS, END_POS, optimized))
  {
    std::cout << "STOMP succeeded" << std::endl;
  }
  else
  {
    std::cout << "A valid solution was not found" << std::endl;
    return -1;
  }
  //! [Solve]

  /**< Further verifying the results */
  if (compareDiff(optimized, trajectory_bias, BIAS_THRESHOLD))
  {
    std::cout << "The solution is within the expected thresholds" << std::endl;
  }
  else
  {
    std::cout << "The solution exceeded the required thresholds" << std::endl;
    return -1;
  }

  return 0;
}
